File: consorcio/lib/domain/usecases/login_use_case.dart
import '../entities/app_user.dart';
import '../errors/domain_exception.dart';
import '../repositories/auth_repository.dart';
import '../repositories/user_repository.dart';
import '../value_objects/user_role.dart';

class LoginUseCase {
  LoginUseCase(this._authRepository, this._userRepository);

  final AuthRepository _authRepository;
  final UserRepository _userRepository;

  static bool _isCredentialError(String message) {
    return message.contains('incorrectos') ||
        message.contains('incorrecto');
  }

  Future<AppUser> execute({
    required String identifier,
    required String password,
  }) async {
    final raw = identifier.trim();
    if (raw.isEmpty || password.isEmpty) {
      throw DomainException('Código y contraseña son obligatorios');
    }

    final dni = raw.contains('@')
        ? null
        : raw.replaceAll(RegExp(r'\D'), '');
    if (dni != null && !RegExp(r'^\d{8}$').hasMatch(dni)) {
      throw DomainException('Ingresa tu código (DNI de 8 dígitos)');
    }

    final email = raw.contains('@')
        ? raw.toLowerCase()
        : await _authRepository.resolveEmailByDni(dni!);

    String userId;
    try {
      userId = await _authRepository.login(email: email, password: password);
    } on DomainException catch (error) {
      // Caché DNI desactualizada tras remapear email en consola.
      if (dni != null && _isCredentialError(error.message)) {
        await _authRepository.clearDniEmailCache(dni);
        final freshEmail = await _authRepository.resolveEmailByDni(
          dni,
          forceRefresh: true,
        );
        if (freshEmail != email) {
          userId = await _authRepository.login(
            email: freshEmail,
            password: password,
          );
        } else {
          rethrow;
        }
      } else {
        rethrow;
      }
    }

    late final AppUser? user;
    try {
      user = await _userRepository.getById(userId);
    } on DomainException catch (error) {
      await _authRepository.logout();
      throw DomainException(error.message);
    } catch (_) {
      await _authRepository.logout();
      throw DomainException(
        'Red lenta al leer tu perfil. Intenta de nuevo.',
      );
    }

    if (user == null) {
      await _authRepository.logout();
      throw DomainException('Usuario sin perfil registrado');
    }

    if (!user.active) {
      await _authRepository.logout();
      throw DomainException('Tu cuenta está desactivada');
    }

    final mobile = user.mobileRoles;
    if (mobile.isEmpty) {
      await _authRepository.logout();
      if (user.assignedRoles.contains(UserRole.superAdministrador)) {
        throw DomainException(
          'El Super Administrador ingresa desde el panel web.',
        );
      }
      throw DomainException(
        'Esta app es para Administrador y Técnico. Usa el panel web.',
      );
    }

    return user;
  }
}

File: src/domain/usecases/auth/LoginUseCase.ts
import type { AuthCredentials, AuthRepository } from '@/domain/repositories/AuthRepository'
import type { UserRepository } from '@/domain/repositories/UserRepository'
import type { User } from '@/domain/entities/User'
import { UnauthorizedError, ValidationError } from '@/domain/errors/DomainError'
import { digitsOnly, DNI_PATTERN } from '@/domain/value-objects/Dni'

export class LoginUseCase {
  private readonly authRepository: AuthRepository
  private readonly userRepository: UserRepository

  constructor(authRepository: AuthRepository, userRepository: UserRepository) {
    this.authRepository = authRepository
    this.userRepository = userRepository
  }

  async execute(credentials: AuthCredentials): Promise<User> {
    const identifier = credentials.identifier.trim()
    const password = credentials.password

    if (!identifier || !password) {
      throw new ValidationError('Correo o código (DNI) y contraseña son obligatorios')
    }

    const email = identifier.includes('@')
      ? identifier.toLowerCase()
      : await this.resolveDniEmail(identifier)

    const userId = await this.authRepository.login(email, password)
    const user = await this.userRepository.getById(userId)

    if (!user) {
      await this.authRepository.logout()
      throw new UnauthorizedError('Usuario sin perfil registrado en el sistema')
    }

    if (!user.active) {
      await this.authRepository.logout()
      throw new UnauthorizedError('Tu cuenta está desactivada. Contacta al administrador')
    }

    return user
  }

  private async resolveDniEmail(identifier: string): Promise<string> {
    const dni = digitsOnly(identifier)
    if (!DNI_PATTERN.test(dni)) {
      throw new ValidationError(
        'Ingresa tu correo electrónico o tu código (DNI de 8 dígitos)',
      )
    }
    return this.authRepository.resolveEmailByDni(dni)
  }
}
